### DISCUSSION TITLE
AIdeck Camera frames streaming crashes

### DISCUSSION THREAD

**Discussion by Saadalh** [bitcraze/discussions#622] [Q&A]
Hello everyone, 

I tried to capture images from the Himax camera attached to an AIdeck using the code provided in 'aideck-gap8-examples/examples/other/wifi-img-streamer/single-shot-capture.py'. I edited the code such that I capture 5 frames, with some time delay between them. However the code crashes after the third capture.

 I figured, since the video streaming code provided in 'aideck-gap8-examples/examples/other/wifi-img-streamer/opencv-viewer.py'  was working fine, I could save the frames that I need from the video stream by tagging each capture with a timestamp and then extracting the captures I needed with the help of these timestamps. However, the code still crashes after approx. 90 frames. 

In both cases, the code gets stuck in the while loop of the 'rx_bytes' function, as the len(data) is 0; meaning that the 'bytearray()' method in line 36 is returning an empty array.

I am new to these topics in general, and would appreciate some guidance on what could cause such an error, and any fix suggestions.

I have attached the code I am running for reference.
[cap.txt](https://github.com/bitcraze/discussions/files/11003384/cap.txt)

Thanks in advance :)
Saad

**Comment by jackyyy65** [comment]
This can cause the stream to become overloaded and crash the code.

One possible solution could be to adjust the time delay between captures to a shorter duration or decrease the number of captures. You could also try increasing the buffer size of the camera stream to prevent it from becoming overloaded.

Additionally, you could try optimizing the c[o](https://thesnowdaypredictor.com/)de to reduce the amount of data being transmitted and processed, which could help improve the stability of the camera stream.

It may also be helpful to review the documentation and resources provided by the manufacturer or community to troubleshoot any issues related to camera streaming and data transmission.

**Reply by hmllr** [reply]
This answer seems to be generated with a tool like ChatGPT - even though there might be some truth in it, those answers will not help much in solving the problem. Please do not answer in our discussions with ChatGPT, it will just generate noise and make the real and useful answers harder to find.

**Comment by hmllr** [comment]
Hi Saad,
this could be an issue at multiple places - either in the python script, but even more likely in the streaming or the camera itself.
To figure this out, I have some more questions:
- is this also happening to you with the original script? Like if you don't save the images, does it also stop streaming after about the same time?
- is your battery fully charged? Are you actually flying while acquisition?
- what is the cfclient console telling you when you don't get images anymore?
There are some known issues with the reliability of the camera, but stopping always after 90 frames is not one of them as far as I know.
Looking forward to more details,
Hanna

**Reply by Saadalh** [reply]
Hello Hanna,

Thank you for your reply. To answer your questions:
- No, the original script runs fine even when I enable the saving option. Some minimal delays are present, but that's about it.
- The battery plays a huge impact, and I always make sure to have at least 75% when testing. The drone is not flying, I attached to a robotic arm to simulate flying instead. 
- I cannot access the cfclient during acquisition, since I am using the radio channel to log IMU pose values.

To explain my specific case more, I am trying to run the video streaming loop in parallel to my python program which executes other commands related to IMU data logging, camera calibration, and robot arm control. Therefore, I had to run the video streaming in a separate python thread. Additionally, in order to stop the stream at some point, I placed the stream loop in a method. This method, along with another method that stops the streaming are placed in a "Camera" class. A copy of this python script is attached to this reply. 

Furthermore, I am initializing a "Camera" object and a thread for each method from another script. Here is where things get laggy and where the data transfer get stuck.

Best regards,
Saad 
[capture.txt](https://github.com/bitcraze/discussions/files/11020722/capture.txt)

**Reply by Saadalh** [reply]
Also worth to mention, sometimes I do get lucky and the stream works properly. It is one time out of many :/

**Reply by hmllr** [reply]
Hmm. I am actually not sure. You not receiving bytes points to no bytes being sent, hence an issue on the esp32 or the gap8/camera. Which is a bit weird, as the way I understood it you did not change this code at all. So the only thing I can think off that changes would be that you are reading it too slow, so you get a buffer overflow somewhere - but I never saw this being the issue before and cannot really think of a place where this should happen. One question here maybe - if you restart the python script but not the drone, does it work again? Also, if you run your code but without the saving, does it still get stuck? 
Another thing to try, just because it is the most common issue I saw with the camera, is to put in a retry in the gap8 code, because under certain circumstances the camera acquisition seems to fail and then block everything (however, I usually only saw this when the cluster is used which is not the case in the wifi streamer). I did that in the face detection streamer: https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/image_processing/FaceDetection/main.c#L363 it is not pretty but helped me to get reliable streaming. I did not try to save the images though.

I hope this helps,
Hanna

**Reply by Saadalh** [reply]
It works randomly. When the battery is charged enough, the camera streams effectively at sometimes only. When it crashes, I try restart the python script, which does the trick sometimes only. If it doesn't, I restart the drone itself and re-establish the wifi connection, sometimes I have to even restart the drone more than once to be able to stream again. 

For the means of my research it is working good enough regardless of the limitations, since I just need stream images and read IMU values to process them further. Thank you for the assistance Hanna :)

**Reply by hmllr** [reply]
Hmm. Non-deterministic errors are hard to debug, especially remotely... It is hard to know where this exactly goes wrong, there are so many microcontrollers involved. I link here two issues for better error signaling on the esp and gap8: https://github.com/bitcraze/aideck-gap8-examples/issues/125 and https://github.com/bitcraze/aideck-esp-firmware/issues/21 - the problem being that if there is an issue with CPX we cannot log the error over CPX anymore and don't know what is wrong.

**Comment by dhamidou** [comment]
Hi, I'm currently experiencing the same problem and would like to know if you managed to resolve the issue.

**Reply by Saadalh** [reply]
Hey Hamidou, unfortunately not. However, operating the drone on a full battery was greatly beneficial (having 2 batteries at hand helps a lot).

**Reply by dhamidou** [reply]
hum.... Thank you for responding so quickly

**Reply by MINJEEgi** [reply]
Hi, Hamidou and Saad. Does it appear that the "single-shot-capture.py" file currently does not exist? Has there been a failure in receiving image data on a timestamp basis?

**Comment by AfrifaEben7** [comment]
Hello,

I am currently facing a similar error running the 'opebcv-viewer.py' example, and it keeps crashing.
Even on full battery, I still experience the same. It worked once, and that is it. Anytime it crashes, I have to restart the drone before I can get some frames, and then it sometimes fails after receiving two frames or more.

I did not make any changes to the original code just added logging and this is what happens;

`[10:25:21] INFO: Stream Session Started`
`[10:25:21] INFO: Log file: logs/stream_20250516_102521.log`
`[10:25:21] INFO: Connecting to socket on 192.168.4.1:5000...`
`[10:25:21] INFO: Socket connected`
`[10:25:21] INFO: Received image: 324x244, depth=1, format=0, size=79056 bytes`
`[10:25:21] INFO: Frame 1: 0.000 s/frame, 2369.66 FPS`
`[10:25:22] INFO: Received image: 324x244, depth=1, format=0, size=79056 bytes`
`[10:25:22] INFO: Frame 2: 0.101 s/frame, 9.89 FPS`
`[10:25:22] INFO: Received image: 324x244, depth=1, format=0, size=79056 bytes`
`[10:25:22] INFO: Frame 3: 0.073 s/frame, 13.68 FPS`
`[10:25:22] INFO: Received image: 324x244, depth=1, format=0, size=79056 bytes`
`[10:26:11] INFO: Stream terminated by user`
`[10:26:11] INFO: Session ended`

**Reply by ToveRumar** [reply]
Hi @AfrifaEben7 !

Im sorry you are experiencing issues with the AI-deck!

Try updating the FW on the AI deck to the latest released version. This can be done by updating the Crazyflie with the latest release in the cfclient, with the AI-deck attached.

This might solve the issue, since we did some improvements on the streaming. If this does not help, please supply the logs from the Crazyflie as well. This will make it easier to understand the issue you are facing. Are you receiving some images and then it stops, or are you unable to receive any image?

**Comment by AfrifaEben7** [comment]
Yes, I receive some frames, then the window freezes then it stops.

`SYS: Crazyflie 2.0 is up and running!`
`SYS: Build 23:5e6a9f70f781 (2025.02 +23) CLEAN`
`SYS: I am 0x333530303536471000390033 and I have 1024KB of flash!`
`CFGBLK: v1, verification [OK]`
`DECK_CORE: 2 deck(s) found`
`DECK_CORE: Calling INIT on driver bcFlow2 for deck 0`
`ZR2: Z-down sensor [OK]`
`PMW: Motion chip id: 0x49:0xB6`
`DECK_CORE: Calling INIT on driver bcAI for deck 1`
`CPX-EXT-ROUTER: Waiting for CPX External router initialization`
`CPX-EXT-ROUTER: CPX External router initialized, CPX_VERSION: 0`
`AIDECK: Not setting up WiFi`
`MPU9250 I2C connection [OK].`
`AK8963 I2C connection [OK].`
`LPS25H I2C connection [OK].`
`ESTIMATOR: Using Kalman (2) estimator`
`CONTROLLER: Using PID (1) controller`
`MTR-DRV: Using brushed motor driver`
`SYS: About to run tests in system.c.`
`EEPROM: I2C connection [OK].`
`STORAGE: Storage check [OK].`
`MPU6500: Self test gyro X [FAIL]. low: -14.0, high: 14.0, measured: -76.43`
`MPU6500: Self test gyro X [FAIL]. low: -14.0, high: 14.0, measured: -30.46`
`MPU6500: Self test gyro X [FAIL]. low: -14.0, high: 14.0, measured: -24.07`
`AK8963: Self test [OK].`
`DECK_CORE: Deck 0 test [OK].`
`DECK_CORE: Deck 1 test [OK].`
S`YS: Self test passed!`
`STAB: Wait for sensor calibration...`
`SYS: Free heap: 9144 bytes`
`CPX: ESP32: I (438) SYS: Initialized`
`CPX: ESP32: I (438) DISCOVERY: Hostname is aideck-7A4228`
`CPX: ESP32: I (438) DISCOVERY: MAC (STA) is 78:21:84:7a:42:28`
`CPX: GAP8: -- WiFi image streamer example --`
`CPX: GAP8: Setting up WiFi AP`
`CPX: GAP8: Starting camera task...`
`CPX: ESP32: I (3178) WIFI: wifi_init_softap finished`
`STAB: Starting stabilizer loop`
`ESTKALMAN: State out of bounds, resetting`
`STAB: WARNING: stabilizer loop rate is off (980)`


@ToveRumar

**Reply by ToveRumar** [reply]
Hi @Saadalh !

Did you try with the latest FW?

Unfortunately, I have seen this issue from time to time. 
Some things you could try is to try is to reduce surrounding radio traffic. Could you go somewhere where the radio traffic might be less intense?
You could also set it up as a wifi station using the menuconfig in the crazyflie fw. 
I have also seen people reducing the frame rate in the GAP8 by introducing a delay before image capture.

This might help at least mitigating the issue. 
Please try these actions and see if it helps!

**Reply by develamit** [reply]
Hi @ToveRumar, 
Stuck on the same wifi image streaming issue. Currently, the wifi is configured in AP mode. The streaming is getting stuck after a few frames. Can you provide an example of how to introduce a delay or increase the buffer size?
"_I have also seen people reducing the frame rate in the GAP8 by introducing a delay before image capture_."

Should it be in this file: https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/other/wifi-img-streamer/wifi-img-streamer.c?
Thanks.

**Reply by IVDC-IITI** [reply]
Even I am facing the same issue, the code crashes after sometime even for the regular wifi streamer provided in the example:
```
...
0.11146472860699051
8.971447851686467
0.1114607068741358
8.97177156008193
0.11146824825100782
8.971164575477742
0.11147184647325324
8.970874993444571
0.1114468408353401
8.972887813639105
^CTraceback (most recent call last):
  File "/home/nambiar/aideck-gap8-examples/examples/other/wifi-img-streamer/opencv-viewer.py", line 102, in <module>
    chunk = rx_bytes(length - 2)
  File "/home/nambiar/aideck-gap8-examples/examples/other/wifi-img-streamer/opencv-viewer.py", line 67, in rx_bytes
    data.extend(client_socket.recv(size-len(data)))
KeyboardInterrupt

```
It works once in many attempted times. not sure how to proceed

**Reply by develamit** [reply]
I've noticed another strange issue. Let me try to explain.
1. I took a stab at customizing wifi-img-streamer.c. 
2. Added a simple string ("My custom") print statement change in line #371 in https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/other/wifi-img-streamer/wifi-img-streamer.c 
3. `cpxPrintToConsole(LOG_TO_CRTP, "-- My custom WiFi image streamer example --\n");`
4. Compiled this .c file following the instructions here: https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/examples/wifi-streamer/
5. But since autotiler is broken, I used the bitcraze/aideck docker image to do the custom make and build. Here are the steps:
`$ cd aideck-gap8-examples
$ docker run --rm -v ${PWD}:/module bitcraze/aideck tools/build/make-example examples/other/wifi-img-streamer image
$ cfloader flash examples/other/wifi-img-streamer/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img deck-bcAI:gap8-fw -w radio://0/80/2M`

After flashing, I ran 
`$ cd aideck-gap8-examples/examples/other/wifi-img-streamer
$ python opencv-viewer.py`

and observed the console log using cfclient. 
To my surprise, the custom log message did not print in the console. It still prints the original log message. 

My question is, why can I not flash a custom image? This is a serious limitation, unless I am doing something wrong. Moreover, how can the aideck print the old stuff after getting flashed with the new change. Completely baffled. Is it cached somewhere?

**Reply by gemenerik** [reply]
> I've noticed another strange issue. Let me try to explain.
> 
> 1. I took a stab at customizing wifi-img-streamer.c.
> 2. Added a simple string ("My custom") print statement change in line #371 in https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/other/wifi-img-streamer/wifi-img-streamer.c
> 3. `cpxPrintToConsole(LOG_TO_CRTP, "-- My custom WiFi image streamer example --\n");`
> 4. Compiled this .c file following the instructions here: https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/examples/wifi-streamer/
> 5. But since autotiler is broken, I used the bitcraze/aideck docker image to do the custom make and build. Here are the steps:
>    `$ cd aideck-gap8-examples $ docker run --rm -v ${PWD}:/module bitcraze/aideck tools/build/make-example examples/other/wifi-img-streamer image $ cfloader flash examples/other/wifi-img-streamer/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img deck-bcAI:gap8-fw -w radio://0/80/2M`
> 
> After flashing, I ran `$ cd aideck-gap8-examples/examples/other/wifi-img-streamer $ python opencv-viewer.py`
> 
> and observed the console log using cfclient. To my surprise, the custom log message did not print in the console. It still prints the original log message.
> 
> My question is, why can I not flash a custom image? This is a serious limitation, unless I am doing something wrong. Moreover, how can the aideck print the old stuff after getting flashed with the new change. Completely baffled. Is it cached somewhere?

You should be able to flash a custom image without any issues. It sounds like one of the following may have happened: the image didn’t flash correctly, the wrong image was flashed, or it didn’t compile properly. Could you try going through all steps and share the console logs so we can take a closer look?

Please open a new discussion since this one is about the streamer crashing.

**Reply by gemenerik** [reply]
> It works once in many attempted times. not sure how to proceed

Could you check if you're running the latest firmware? There have been several recent improvements to streaming, so updating may help resolve the issue and make debugging easier on our end.

**Reply by gemenerik** [reply]
> Hi @ToveRumar, Stuck on the same wifi image streaming issue. Currently, the wifi is configured in AP mode. The streaming is getting stuck after a few frames. Can you provide an example of how to introduce a delay or increase the buffer size? "_I have also seen people reducing the frame rate in the GAP8 by introducing a delay before image capture_."
> 
> Should it be in this file: https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/other/wifi-img-streamer/wifi-img-streamer.c? Thanks.

Yes, you can introduce a delay here.

**Reply by IVDC-IITI** [reply]
> > It works once in many attempted times. not sure how to proceed
> 
> Could you check if you're running the latest firmware? There have been several recent improvements to streaming, so updating may help resolve the issue and make debugging easier on our end.

Yes, i have installed the latest firmware. I have tried everything form reinstalling the firmware to using it while completely charged.
I am using one of the examples codes from https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/other/wifi-img-streamer/opencv-viewer.py

**Reply by IVDC-IITI** [reply]
Also when you speak of battery what is the required battery voltage we are expecting here. I have used it when battery was at around 4V and still it freezes after running for a min or so

**Reply by gemenerik** [reply]
Around 4V should be fine. Sorry to hear it is still freezing after you updated. Let's continue the conversation in the linked issue.

**Comment by gemenerik** [comment]
We've opened a dedicated issue to track the AI‑deck streaming freeze problem and consolidate efforts: https://github.com/bitcraze/aideck-gap8-examples/issues/150

**Comment by pistoe** [comment]
Does the aideck-gap8-examples/examples/other/wifi-img-streamer/single-shot-capture.py example still exist somewhere?

**Reply by gemenerik** [reply]
Hi! I don't think that file ever existed? I can find zero trace of it and have no knowledge of it. May be some LLM hallucination.
